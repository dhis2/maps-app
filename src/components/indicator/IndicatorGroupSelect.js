import i18n from '@dhis2/d2-i18n'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { loadIndicatorGroups } from '../../actions/indicators.js'
import { SelectField } from '../core/index.js'

class IndicatorGroupSelect extends Component {
    static propTypes = {
        loadIndicatorGroups: PropTypes.func.isRequired,
        onChange: PropTypes.func.isRequired,
        className: PropTypes.string,
        errorText: PropTypes.string,
        indicatorGroup: PropTypes.shape({
            id: PropTypes.string.isRequired,
        }),
        indicatorGroups: PropTypes.array,
    }

    componentDidMount() {
        const { indicatorGroups, loadIndicatorGroups } = this.props

        if (!indicatorGroups) {
            loadIndicatorGroups()
        }
    }

    render() {
        const {
            indicatorGroups,
            indicatorGroup,
            onChange,
            className,
            errorText,
        } = this.props

        return (
            <SelectField
                label={i18n.t('Indicator group')}
                loading={indicatorGroups ? false : true}
                items={indicatorGroups}
                value={indicatorGroup ? indicatorGroup.id : null}
                onChange={onChange}
                className={className}
                errorText={!indicatorGroup && errorText ? errorText : null}
                dataTest="indicatorgroupselect"
            />
        )
    }
}

export default connect(
    (state) => ({
        indicatorGroups: state.indicatorGroups,
    }),
    { loadIndicatorGroups }
)(IndicatorGroupSelect)
