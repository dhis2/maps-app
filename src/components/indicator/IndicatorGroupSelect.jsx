import { useDataQuery } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import PropTypes from 'prop-types'
import React from 'react'
import { SelectField } from '../core/index.js'

// Load all indicator groups
const INDICATOR_GROUPS_QUERY = {
    groups: {
        resource: 'indicatorGroups',
        params: {
            fields: ['id', 'displayName~rename(name)'],
            paging: false,
        },
    },
}

const IndicatorGroupSelect = ({
    indicatorGroup,
    onChange,
    className,
    errorText,
}) => {
    const { loading, error, data } = useDataQuery(INDICATOR_GROUPS_QUERY)

    return (
        <SelectField
            label={i18n.t('Indicator group')}
            loading={loading}
            items={data?.groups.indicatorGroups}
            value={indicatorGroup?.id}
            onChange={onChange}
            className={className}
            errorText={
                error?.message ||
                (!indicatorGroup && errorText ? errorText : null)
            }
            dataTest="indicatorgroupselect"
        />
    )
}

IndicatorGroupSelect.propTypes = {
    onChange: PropTypes.func.isRequired,
    className: PropTypes.string,
    errorText: PropTypes.string,
    indicatorGroup: PropTypes.shape({
        id: PropTypes.string.isRequired,
    }),
}

export default IndicatorGroupSelect
