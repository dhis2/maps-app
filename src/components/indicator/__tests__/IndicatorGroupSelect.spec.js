import React from 'react';
import { shallow } from 'enzyme';
import { SelectField } from '@dhis2/d2-ui-core';
import { IndicatorGroupSelect } from '../IndicatorGroupSelect';

// TODO:
// clear the spied function after each test run: https://stackoverflow.com/questions/43245040/using-jest-to-spy-on-method-call-in-componentdidmount

describe('IndicatorGroupSelect', () => {
    const renderWithProps = props => shallow(<IndicatorGroupSelect
        {...props}
    />);

    it('should render a d2-ui SelectField', () => {
        const loadIndicatorGroups = jest.fn();
        const onChange = jest.fn();
        expect(renderWithProps({ loadIndicatorGroups, onChange }).type()).toBe(SelectField);
    });

    it('should call loadIndicatorGroups function in componentDidMount if no indicator groups exists', () => {
        const loadIndicatorGroups = jest.fn();
        const onChange = jest.fn();
        const select = renderWithProps({ loadIndicatorGroups, onChange });

        select.instance().componentDidMount();
        expect(loadIndicatorGroups).toHaveBeenCalled();
    });

    it('should not call loadIndicatorGroups function in componentDidMount if indicator groups exists', () => {
        const loadIndicatorGroups = jest.fn();
        const onChange = jest.fn();
        const select = renderWithProps({ loadIndicatorGroups, onChange, indicatorGroups: [] });

        select.instance().componentDidMount();
        expect(loadIndicatorGroups).not.toHaveBeenCalled();
    });

});
