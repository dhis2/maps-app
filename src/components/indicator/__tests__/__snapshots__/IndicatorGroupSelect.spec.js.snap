// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndicatorGroupSelect calls loadIndicatorGroups if no indicator groups exist 1`] = `
<div>
  <div
    class="inputField"
  >
    <div
      class="ui-SingleSelectField"
    />
  </div>
</div>
`;

exports[`IndicatorGroupSelect renders SelectField 1`] = `
<div>
  <div
    class="inputField"
  >
    <div
      class="ui-SingleSelectField"
    >
      <div
        class="ui-SingleSelectOption"
      />
      <div
        class="ui-SingleSelectOption"
      />
    </div>
  </div>
</div>
`;
